### PR TITLE
[5.9] More batched changes

### PR DESCRIPTION
**Pull request by gottesmm**
This PR contains a set of batched commits for 5.9. A few of the commits were
test tweaks so I did not include CCC for them.

Commit 6a5ef8af94e38e720ad0924b4721493a3c575b6f
• Description: [move-only] Make sure that we serialize deinits and if we are not
  able to see the value witness function, call it via the value witness
  function.
• Risk: Low. This only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/66185
• Reviewed By: @jckarter
• Testing: I put in both a swiftpm like test case and a library evolution test
  case. I also updated the moveonly_deinit serialization swift test to show that
  we actually serialize the deinit.
• Resolves: rdar://109679168

Commit 5530aa95151e92f7326b58cadb40252d2dafb8b4

• Description: [move-only] Teach deinit devirtualization how to handle
  devirtualizing a deinit of a resilient type.
• Risk: Low. This only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/66185
• Reviewed By: @jckarter
• Testing: I added tests that exercised this behavior.
• Resolves: rdar://109904633

Commit 45a592c2054d1768c022134e6b77195d6b44e7f2
• Description: [move-only] Change closure capture diagnostic for let patterns to
  say that it cannot be captured by an escaping closure.
• Risk: Low. This only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/66197
• Reviewed By: @jckarter
• Testing: Updated tests to show we emit the correct diagnostic.
• Resolves: rdar://109742587

Commit 20bf5d58f1d7d76be0da2e82222794861e43c322
• Description: [move-only] Teach SILGenApply how to emit subscripts with borrowed base values.
• Risk: Low. This only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/66094
• Reviewed By: @jckarter
• Testing: Added exhaustive tests in this area to test out the behavior
• Resolves: rdar://109746476

Commit c009c324ae7941950fb7b82053da33fa48980b1b
• Description: [move-only] When we emit a noncopyable temporary allocation,
  always attach a mark_must_check [consumable_and_assignable]. This removes most
  of the copy of noncopyable type errors seen in the tests from
  20bf5d58f1d7d76be0da2e82222794861e43c322
• Risk: Low. This only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/66094
• Reviewed By: @jckarter
• Testing: Added exhaustive tests in this area to test out the behavior
• Resolves: rdar://109746476

Commit 61745eedee810e9b8abd02be08cbb95ef7bf0ec3
• Description: [move-only] Make sure that we handle load [take] that are
  actually borrows correctly.
• Risk: Low. This only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/66094
• Reviewed By: @jckarter
• Testing: Added SIL tests that show that we have the correct behavior here
• Resolves: rdar://109746476

Commit 23deccfb6ca090be7fc6588bca6b144f62207b16
• Description: Restricts c009c324ae7941950fb7b82053da33fa48980b1b so we only
  perform the transformation on reads instead of also gets.
• Risk: Low. This only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/66094
• Reviewed By: @jckarter
• Testing: Updated silgen tests that show this behavior. Updated the subscript
  tests so that they validate that we do not have copy of noncopyable errors in
  the places that this change fixes.
• Resolves: rdar://109746476

Commit 1ae66b2fd583b10b9d13dceca0a156d25f3118b7
• Description: [move-only] Teach SILGen how to recognize a borrowed read
  subscript when the type also has a modify.
• Risk: Low. This only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/66094
• Reviewed By: @jckarter
• Testing: Updated silgen tests that show this behavior. Updated the subscript
  tests so that they validate that we do not have copy of noncopyable errors in
  the places that this change fixes.
• Resolves: rdar://109746476

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

File: lib/IRGen/GenType.cpp [hunk 1]
```diff
@@ -2887,19 +2887,25 @@ static bool tryEmitDeinitCall(IRGenFunction &IGF,
...
-  
-  auto deinit = IGF.getSILModule().lookUpMoveOnlyDeinit(nominal);
-  assert(deinit && "type has a deinit declared in AST but SIL deinit record is not present!");
-    
+
+  auto deinitTable = IGF.getSILModule().lookUpMoveOnlyDeinit(nominal);
+
+  // If we do not have a deinit table, call the value witness instead.
+  if (!deinitTable) {
+    irgen::emitDestroyCall(IGF, T, indirect());
```
Comment (jckarter) [inline lib/IRGen/GenType.cpp:2900]: You need to call `indirectCleanup` after the call to `indirect`.

Comment (gottesmm) [reply lib/IRGen/GenType.cpp:2900]: I will prepare a small patch on main that will fix this. Why don't you give the +1 and then I can get TimK to +1 this and I can take care of this tonight/over the weekend.

**Review by jckarter — APPROVED** [review]
This should be safe to land now without the `indirectCleanup` change, which is minor enough to do in a follow-up commit.

**Review by tbkka — APPROVED** [review]

**Comment by gottesmm** [comment]
@swift-ci test

**Pull request merged**